content/blog/gatbsy-typescript/index.md

```diff
@@ -0,0 +1,284 @@
+and why I do think that most medium to largse size Javascript applications
+would benefit from using Typescript, in no particular order:
+
+1- automatically up-to-date and synchronized code documentation in the form of types.
+2- new code verification tool added to your pool (in javascript apps this is mostly made of different kinds of tests and lints)
+3- better development experience: better autocompletion and api discovery
+4- type checker guided hassle free refactoring
+5- improve the mind set of developers so that they think more about data structures and their semantic meaning inside the code (as oposed to just have a bunch of unamed and uncategorized objects which is something that tends to happen in bigger javascript apps).
+
+I plan to talk a lot more about this in a follow up post so that is why I wont focus to much
+explaining these points, but please note that most of these come from real world experiences
+handling large production code bases in both Javascript and Typescript, and there is an abysm between the too.
+
+Additionally, from a business standpoint we can also enumerate derived benefits:
+
+- more productivity: items 1, 3 and 4
+- higher quality: items 2, 4 and 5
+
+Having these in mind we are going to cover how to integrate Typescript to Gatsby,
+considering that Gatbsy is a framework that can accomodote larger production applications that
+will benefit the most out of this.
+
+The process is rather simple but I want to cover some paint points and tips and tricks.
+
+Note: This blog is actually using Typescript.
+
+## How to use Typescript in your Gatbsy.js site.
+
+We are going to install two plugins:
+
+- [gatsby-plugin-typescript](https://www.gatsbyjs.org/packages/gatsby-plugin-typescript/) to support typescript compilation
+- [gatsby-plugin-graphql-codegen](https://www.gatsbyjs.org/packages/gatsby-plugin-graphql-codegen/) to automatically generate Typescript types out of the underlying graphql data model that Gatsby uses.
+
+After following the instructions you will have most of the tools, the rest is going to be
+related to 1) configuration and 2) migrating js files.
+
+## Configuration
+
+### Typescript
+
+You will need to create a `tsconfig.json` at the root of your project configuring the
+typescript compiler. This can vary a lot depending on your taste but here's the one we are using:
+
+```json
+{
+  "compilerOptions": {
+    "target": "esnext",
+    "module": "commonjs",
+    "jsx": "preserve",
+    "strict": true,
+    "noImplicitAny": true,
+    "strictNullChecks": false,
+    "strictFunctionTypes": true,
+    "noImplicitThis": true,
+    "alwaysStrict": true,
+    "noUnusedLocals": true,
+    "noUnusedParameters": true,
+    "noImplicitReturns": true,
+    "noFallthroughCasesInSwitch": true,
+    "allowSyntheticDefaultImports": true,
+    "esModuleInterop": true
+  },
+  "include": ["src/**/*"]
+}
+```
+
+I wont cover all the options but I do want to cover `strictNullChecks`.
```

Comment:
won't

Reply:
I also think it's a good idea to explain what `strictNullChecks` does before going on to explain how it was problematic.

Reply:
done!